### PR TITLE
Add 404 page

### PR DESCRIPTION
**Pull request by paulbert**
Adds a 404.md and updates the mkwiki html (new index.html) to use that file when a page isn't found.

See issue https://github.com/open-learning-exchange/open-learning-exchange.github.io/issues/479

Also, you can preview 404 functionality [here](https://rawgit.com/paulbert/paulbert.github.io/add_404/#!pages/firststps.md)

### CODE REVIEW

**Comment by paulbert** [comment]
This PR is now a bit intertwined with this one: https://github.com/open-learning-exchange/open-learning-exchange.github.io/pull/483

The short version is that if that change is approved the index.html from that PR should be merged and this one discarded, but if that one is not I'll need to update the index.html on this PR before merging.

The long version is that I rebuilt the index.html from mdwiki before seeing that there were a few scripts to add after getting the latest version.  I included those scripts in a [new fork of mdwiki](https://github.com/paulbert/mdwiki/tree/0.6.x) and rebuilt the index.html for that PR, but that also includes the change for the offline version per that PR.  By its nature as the latest 0.6.x version of mdwiki it also includes the 404 support, so if that PR is merged this index.html is not needed.

**Comment by sente** [comment]
Can you also squash/rebase your commits before merging?

**Comment by paulbert** [comment]
All squashed.

The index.html included here is the slim version with the modifications for the OLE site.  The offline PR #483 will add the inline HTML js & css libraries.

**Comment by dogi** [comment]
@dogi is confused

@paulbert what is the difference to https://github.com/open-learning-exchange/open-learning-exchange.github.io/pull/483 ?

**Comment by paulbert** [comment]
@dogi @mappuji The goal of this PR is adding a 404 page and functionality to the wiki.  It adds a 404.md and a newer version of the mdwiki html which will use the 404.md file if the file requested is not found.  You can see that it is working by trying to load an md file that doesn't exist [like this one](https://rawgit.com/paulbert/paulbert.github.io/add_404/#!pages/firststps.md) to see that it uses the 404.md.  It differs from #483 in its intended goal: that PR is intended simply to get the wiki working offline when loaded from the filesystem (file:///).

The two PRs are similar because both led me to creating [this fork](https://github.com/paulbert/mdwiki/tree/0.6.x) of mdwiki to incorporate the smaller changes that have accumulated to the index.html into the build process rather than added post build to the html output.  Because of that some of the changes that are specifically for the offline PR made it into this one, specifically the change for the Bootstrap theme selector and the additional styles for the offline font.  Because of this both PRs would benefit from a bit of scrutiny from those of you more familiar with the wiki to make sure there aren't additional customizations to the index.html missing.

Let me know if you need more clarity on this.

File: index.html [hunk 1]
```diff
@@ -173,38 +208,17 @@
...
...em,"gm")(/bull/g,i.bullet)(),i.list=e(i.list)(/bull/g,i.bullet)("hr",/\n+(?=(?: *[-*_]){3,} *(?:\n+|$))/)(),i._tag="(?!(?:a|em|strong|small|s|cite|q|dfn|abbr|data|time|code|var|samp|kbd|sub|sup|i|b|u|mark|ruby|rt|rp|bdi|bdo|span|br|wbr|ins|del|img)\\b)\\w+(?!:/|@)\\b",i.html=e(i.html)("comment",/<!--[\s\S]*?-->/)("closed",/<(tag)[\s\S]+?<\/\1>/)("closing",/<tag(?:"[^"]*"|'[^']*'|[^'">])*?>/)(/tag/g,i._tag)(),i.paragraph=e(i.paragraph)("hr",i.hr)("heading",i.heading)("lheading",i.lheading)("blockquote",i.blockquote)("tag","<"+i._tag)("def",i.def)(),i.normal=g({},i),i.gfm=g({},i.normal,{fences:/^ *(`{3,}|~{3,}) *(\S+)? *\n([\s\S]+?)\s*\1 *(?:\n+|$)/,paragraph:/^/}),i.gfm.paragraph=e(i.paragraph)("(?!","(?!"+i.gfm.fences.source.replace("\\1","\\2")+"|")(),i.tables=g({},i.gfm,{nptable:/^ *(\S.*\|.*)\n *([-:]+ *\|[-| :]*)\n((?:.*\|.*(?:\n|$))*)\n*/,table:/^ *\|(.+)\n *\|( *[-:]+[-| :]*)\n((?: *\|.*(?:\n|$))*)\n*/}),a.rules=i,a.lex=function(b,c){var d=new a(c);return d.lex(b)},a.prototype.lex=function(a){return a=a.replace(/\r\n|\r/g,"\n").replace(/\t/g,"    ").replace(/\u00a0/g," ").replace(/\u2424/g,"\n"),this.token(a,!0)},a.prototype.token=function(a,b){for(var c,d,e,f,g,h,j,k,l,a=a.r
```
Comment (dogi) [inline index.html]: @paulbert do we have to delete this lines?

I think this also was added some virtual interns some time ago ;)

File: index.html [hunk 2]
```diff
@@ -86,9 +86,44 @@
         font-size: smaller;
         padding: 1em;
     }
-	.navbar-brand {
-		margin-left: 35px;
+	
+	/* Begin Added for Open Learning Exchange */
```
Comment (dogi) [inline index.html]: @paulbert yes we added also this 3 lines - good job

File: index.html (hunk 1 above)
Comment (dogi) [inline index.html]: @paulbert this is a very big change - wonder can you explain what changed? aslo wonder about the source of your change?

Comment (dogi) [inline index.html]: @paulbert this is an unnecessary change - please undo

Comment (dogi) [inline index.html]: something with this line preview is wrong I mean this lines of the original index.html

```
<script type='text/javascript'>
 - $(document).ready(function() { 
 -    $( "body" ).on( "click", "a", function() {
 -        $("a[href^='http://']").each(
 -        function(){
 -            if(this.href.indexOf(location.hostname) == -1) {
 -                $(this).attr('target', '_blank');
 -            }
 -    });
 -         
 -    $("a[href^='https://']").each(
 -        function(){
 -            if(this.href.indexOf(location.hostname) == -1) {
 -                $(this).attr('target', '_blank');
 -            }
 -    });
 -});
 - });
 -</script>
```

Comment (dogi) [inline index.html]: also here same as above - on both I had added my comment to the deleted original lines maybe that is what is confusing github ...

anyhow I mean this line
```
<a href="http://ole.org"><img style="z-index:2000;width: 35px;position: fixed;margin-top: 12.5px;margin-left: 9px;" src="pages/uploads/images/logo.png"></a>
```

Comment (paulbert) [inline index.html]: @dogi The big block of code that's changing is the minified mdwiki js which is part of the update to use the 404 page.  It includes a very small change that they had made to add support for a 404.md, in addition to some changes to add some of the custom functionality that needed to be added manually previously including the script to add `target="_blank"` to all external links.  Here's a quick list of what is now part of the custom mdwiki build process on the [fork I made here:](https://github.com/paulbert/mdwiki/tree/0.6.x)

- The flatly theme for Bootstrap is the main theme rather than the default, allowing the flatly theme to be used offline. The default Bootstrap theme can be chosen in navigation.md with 'classic' (this is the change which is specifically for PR #483 rather than this one)
- Added a function to basic_skeleton.js which adds target="_blank" to any links to external domains
- Adds a line to the setPageTitle function which will set the text of the title tag to the text of the opening h1 tag of the page if found.
- Adds OLE logo in upper left.
- Includes added styles for offline use of the Lato font.
- Changes favicon from a favicon.png to favicon.ico file in the base directory.

Comment (paulbert) [inline index.html]: @dogi To add a little clarity: this script is in the minified mdwiki code.  External links still open a new tab in the browser when I check it [here](https://rawgit.com/paulbert/paulbert.github.io/add_404/#!index.md).  Please let me know if there's a particular external link that is not working and I can try to replicate it on my end to fix it.

Comment (dogi) [inline index.html]: Thanks that explains it ;)

Comment (dogi) [inline index.html]: in https://rawgit.com/paulbert/paulbert.github.io/add_404/#!index.md it seems to work ...
